### PR TITLE
ESQL: Remove the `NESTED` DataType

### PR DESCRIPTION
**Pull request by nik9000**
We don't support nested fields at the moment and when we do I'm not sure it'll *be* a DataType. Maybe. Probably. But let's deal with that when we support it.

### CODE REVIEW

File: x-pack/plugin/esql/qa/testFixtures/src/main/java/org/elasticsearch/xpack/esql/LoadMapping.java [hunk 1]
```diff
@@ -82,7 +81,7 @@ private static void walkMapping(String name, Object value, Map<String, EsField>
             // extract field type
             DataType esDataType = getType(content);
             final Map<String, EsField> properties;
-            if (esDataType == OBJECT || esDataType == NESTED) {
+            if (esDataType == OBJECT) {
```
Comment (nik9000) [inline x-pack/plugin/esql/qa/testFixtures/src/main/java/org/elasticsearch/xpack/esql/LoadMapping.java:84]: This is from `Types` in QL, but it's only used in testing in ESQL. I think this safe for now.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Analyzer.java [hunk 2]
```diff
@@ -245,8 +244,8 @@ private static void mappingAsAttributes(List<Attribute> list, Source source, Fie
                 if (DataType.isPrimitive(type)) {
                     list.add(attribute);
                 }
-                // allow compound object even if they are unknown (but not NESTED)
-                if (type != NESTED && fieldProperties.isEmpty() == false) {
+                // allow compound object even if they are unknown
+                if (fieldProperties.isEmpty() == false) {
```
Comment (nik9000) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Analyzer.java:248]: This is the most interesting line, I think. But we don't *make* `NESTED` any more so I feel like this is pretty safe.

**Comment by nik9000** [comment]
I talked with @astefan and I'm going to make a yaml test that uses nested just to make sure we're not doing something really unexpected. It *should* be fine, but let's be super careful here.

**Comment by nik9000** [comment]
Mechanically what happens for `nested` fields and all fields under them is that the `IndexResolver` hits the `nested` field and marks it as unsupported. Then it hits the children of the `nested` field and marks them as unsupported *because* the parent is unsupported.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Analyzer.java (hunk 2 above)
Comment (nik9000) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Analyzer.java:248]: So this *would* change things. It'd make the sub-fields of the `NESTED` object appear. I think. I'll double check.

**Comment by nik9000** [comment]
> Mechanically what happens for `nested` fields and all fields under them is that the `IndexResolver` hits the `nested` field and marks it as unsupported. Then it hits the children of the `nested` field and marks them as unsupported _because_ the parent is unsupported.

We *don't* want this. At least, it's a change and maybe not a good one. `nested` fields aren't the same as non-nested fields. Previously we were removing them entirely from ESQL. It just didn't see them. With this change as it stands now they'd start showing up as `nested`. I don't think we want that, at least, not now. I'm going to try and filter them out earlier.

**Comment by nik9000** [comment]
OK! I've moved the filtering of `nested` fields lower. That should be better....

**Review by astefan — APPROVED** [review]
I think it's ok to remove this data type and I hope we are careful enough with this step.
Maybe add two-three more to `40_unsupported_types.yml` with a mapping that more paranoid:

```
            "name": {
                "type": "object",
                "properties": {
                    "first_name": {
                        "type": "keyword"
                    },
                    "last_name": {
                        "type": "keyword"
                    }
                }
            },
            "bla": {
                "type": "object",
                "properties": {
                    "obj": {
                        "type": "object",
                        "properties": {
                            "first_name": {
                                "type": "keyword"
                            },
                            "last_name": {
                                "type": "keyword"
                            }
                        }
                    }
                }
            },
            "nested": {
                "type": "nested",
                "properties": {
                    "sub_nested": {
                        "type": "nested",
                        "properties": {
                            "first_name": {
                                "type": "keyword"
                            }
                        }
                    },
                    "last_name": {
                        "type": "keyword"
                    }
                }
            }
```

**Comment by nik9000** [comment]
> Maybe add two-three more to `40_unsupported_types.yml` with a mapping that more paranoid:

:+1:

**Comment by nik9000** [comment]
That snapshot failures are real - old versions sometimes send `nested` as the data type:
```
[2024-08-05T12:22:54,015][WARN ][o.e.x.e.a.EsqlResponseListener] [test-cluster-0] Request failed with status [INTERNAL_SERVER_ERROR]: java.io.IOException: Unknown DataType for type name: nested
	at org.elasticsearch.xpack.esql.core.type.DataType.readFrom(DataType.java:449)
	at org.elasticsearch.xpack.esql.core.type.EsField.<init>(EsField.java:57)
	at org.elasticsearch.common.io.stream.NamedWriteableAwareStreamInput.readNamedWriteable(NamedWriteableAwareStreamInput.java:56)
	at org.elasticsearch.common.io.stream.NamedWriteableAwareStreamInput.readNamedWriteable(NamedWriteableAwareStreamInput.java:32)
	at org.elasticsearch.xpack.esql.io.stream.PlanNamedTypes.lambda$readEsIndex$3(PlanNamedTypes.java:771)
	at org.elasticsearch.common.io.stream.StreamInput.readImmutableMap(StreamInput.java:
	at org.elasticsearch.xpack.esql.io.stream.PlanNamedTypes.readEsIndex(PlanNamedTypes.java:771)
	at org.elasticsearch.xpack.esql.io.stream.PlanNamedTypes.readEsRelation(PlanNamedTypes.java:545)
```

When we serialize `EsRelation` we serialize the original `EsIndex` in addition to the `

**Comment by nik9000** [comment]
@astefan I pushed a change just now to explain the reading change better. It's to do with the way we reuse a function for serialization.

**Comment by astefan** [comment]
Thank you @nik9000

**Comment by nik9000** [comment]
> Thank you @nik9000

That sounds like and approal. I'm mash the merge button. Let me know if you have any more ideas around this one. It's been a trip!

**Pull request merged**